### PR TITLE
Assignment3 Chris Cornell

### PR DESCRIPTION
**Pull request by ccornell88**

### CODE REVIEW

**Comment by Oscar-Ruelas** [comment]
Assignment
For this assignment do the following:

Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
Fundamentals
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on treat warnings as errors for all projects ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
For this assignment, always use Assert.Equal<T>() (the generic version) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Comment by Oscar-Ruelas** [comment]
Nice Job, Chris I couldn't find any flaws with your code ;). Perhaps you could try to implement the extra credit. It's some very easy points to get.

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 1]
```diff
@@ -1,7 +1,89 @@
...
+
+        //Assert
+        outputMock.Verify(foo => foo.WriteJoke("Why is 6 afraid of 7? Because 7 ate 9!"), Times.Once());
+    }
+
+    [Fact]
+    public void TellJoke_ChuckNorrisJokeRetrieved_NewJokeChosenAndOutputCorrectly()
+    {
+        //Arrange
+        var serviceMock = new Mock<IJokeService>();
```
Comment (theRoice) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:30]: Reading these lines makes me think we implemented our mock objects incorrectly and helped me understand how the moq library applies to our project a little bit better!

**Comment by theRoice** [comment]
Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
Fundamentals
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on treat warnings as errors for all projects ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
For this assignment, always use Assert.Equal() (the generic version) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Comment by theRoice** [comment]
Code looks solid and makes me want to give our code a once over to clean it up a bit more. If I could offer one suggestion it would be to consider using file-scoped namespaces (putting semi-colons after your namespaces) in order to eliminate the first pair of braces and one level of indentation from your code. Both ways are valid but I figured I might offer an alternative!

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 2]
```diff
@@ -1,7 +1,89 @@
...
+    [Fact]
+    public void OutputService_Initialized_ReturnsSameInstance()
+    {
+        // Arrange
+        var mockJokeService = new Mock<IJokeService>();
+        var mockOutputService = new Mock<IOutputService>();
+        var jester = new Jester(mockOutputService.Object, mockJokeService.Object);
+
+        // Act
+        var result = jester.OutputService;
```
Comment (AndrewKorchemniy) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:69]: The dependencies taken in by Jester should be private.

File: CanHazFunny/CanHazFunny.Tests/JokeServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            // Arrange
+            var jokeServiceMock = new Mock<IJokeService>();
+            string expectedJoke = "What do you call fake spaghetti? An impasta!";
+            jokeServiceMock.Setup(service => service.GetJoke()).Returns(expectedJoke);
+
+            // Act
+            string actualJoke = jokeServiceMock.Object.GetJoke();
+
+            // Assert
+            Assert.Equal(expectedJoke, actualJoke);
```
Comment (AndrewKorchemniy) [inline CanHazFunny/CanHazFunny.Tests/JokeServiceTests.cs:21]: This is testing the mock and not the implementation.

File: CanHazFunny/CanHazFunny.Tests/JokeServiceTests.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        [Fact]
+        public void GetJoke_Called_ReturnsNonEmptyString()
+        {
+            // Arrange
+            var service = new JokeService();
```
Comment (AndrewKorchemniy) [inline CanHazFunny/CanHazFunny.Tests/JokeServiceTests.cs:28]: We aren't passing in mocked dependencies here - not sure what this is testing.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+﻿using System;
+
+namespace CanHazFunny
+{
+    public class Jester
+    {
+
+        public IJokeService JokeService { get; }
+        public IOutputService OutputService { get; }
```
Comment (AndrewKorchemniy) [inline CanHazFunny/CanHazFunny/Jester.cs:9]: These should be private.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    {
+
+        public IJokeService JokeService { get; }
+        public IOutputService OutputService { get; }
+
+        public Jester(IOutputService? outputService, IJokeService? jokeService)
+        {
+            JokeService = jokeService ?? throw new ArgumentNullException(nameof(jokeService));
+            OutputService = outputService ?? throw new ArgumentNullException(nameof(outputService));
+        }
```
Comment (AndrewKorchemniy) [inline CanHazFunny/CanHazFunny/Jester.cs:15]:
```suggestion
        public Jester(IOutputService outputService, IJokeService jokeService)
        {
            JokeService = jokeService ?? throw new ArgumentNullException(nameof(jokeService));
            OutputService = outputService ?? throw new ArgumentNullException(nameof(outputService));
        }
```

Null checks are a valid use-case even on non-nullable members.
https://stevetalkscode.co.uk/null-injection
We shouldn't trick users into passing in nullable services when doing so will throw an exception immediately.

File: CanHazFunny/CanHazFunny/Program.cs [hunk 7]
```diff
@@ -1,10 +1,11 @@
-﻿namespace CanHazFunny;
+﻿using System;
+namespace CanHazFunny;
 
-class Program
+public sealed class Program
```
Comment (AndrewKorchemniy) [inline CanHazFunny/CanHazFunny/Program.cs:4]: Program is rather unlikely to be derived from.

**Pull request closed without merging**